{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b95a0b23",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import libraries\n",
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0d5332db",
   "metadata": {},
   "source": [
    "### VDJdb columns explanation\n",
    "gene column contains TRA and TRB, TRA means TCR alpha subunit, TRB means TCR beta subunit. \n",
    "\n",
    "cdr3 column is the segment to join the TRA and TRB together.  \n",
    "\n",
    "v.segm and j.segm columns is part of TRA or TRB.  \n",
    "\n",
    "sepecies column is represent where is gene from, here is HomoSapiens, also called human(All of us).\n",
    "\n",
    "Before explaning columns, we need to know what is MHC. MHC means Major histocompatibility complex, which contail 2 classes.\n",
    "MHC I class and MHC II class (Here is mhc.class, MHCI and MHCII). Both classes of proteins share the task of presenting peptides on the cell surface for recognition by T cells.There is a set of alleles that will determine the protein type and form of MHCI(Here, mhc.a and mhc.b is expressed from alleles gene regions(Maybe, i am not sure)), more details please click https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5355494/#:~:text=Major%20histocompatibility%20complex%20(MHC)%20class%20I%20and%20class%20II%20proteins,for%20recognition%20by%20T%20cells.\n",
    "\n",
    "mhc.a and mhc.b columns is one gene expression based on mhc.class.  \n",
    "\n",
    "mhc.class column is to specify a class the gene belong to.  \n",
    "\n",
    "antigen.epitope column , also known as antigenic determinant, is the part of an antigen that is recognized by the immune system, specifically by antibodies, B cells, or T cells. The part of an antibody that binds to the epitope is called a paratope.  \n",
    "\n",
    "antigen.species columns is show us where is the paratope or epitope from, here the mhc's paratope from HIV-1.  \n",
    " \n",
    "More columns details, Please click: https://github.com/antigenomics/vdjdb-db"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "118ff026",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>complex.id</th>\n",
       "      <th>gene</th>\n",
       "      <th>cdr3</th>\n",
       "      <th>v.segm</th>\n",
       "      <th>j.segm</th>\n",
       "      <th>species</th>\n",
       "      <th>mhc.a</th>\n",
       "      <th>mhc.b</th>\n",
       "      <th>mhc.class</th>\n",
       "      <th>antigen.epitope</th>\n",
       "      <th>...</th>\n",
       "      <th>antigen.species</th>\n",
       "      <th>reference.id</th>\n",
       "      <th>method</th>\n",
       "      <th>meta</th>\n",
       "      <th>cdr3fix</th>\n",
       "      <th>vdjdb.score</th>\n",
       "      <th>web.method</th>\n",
       "      <th>web.method.seq</th>\n",
       "      <th>web.cdr3fix.nc</th>\n",
       "      <th>web.cdr3fix.unmp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>TRA</td>\n",
       "      <td>CIVRAPGRADMRF</td>\n",
       "      <td>TRAV26-1*01</td>\n",
       "      <td>TRAJ43*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-B*08</td>\n",
       "      <td>B2M</td>\n",
       "      <td>MHCI</td>\n",
       "      <td>FLKEKGGL</td>\n",
       "      <td>...</td>\n",
       "      <td>HIV-1</td>\n",
       "      <td>PMID:15596521</td>\n",
       "      <td>{\"frequency\": \"\", \"identification\": \"tetramer-...</td>\n",
       "      <td>{\"cell.subset\": \"CD8+\", \"clone.id\": \"\", \"donor...</td>\n",
       "      <td>{\"cdr3\": \"CIVRAPGRADMRF\", \"cdr3_old\": \"CIVRAPG...</td>\n",
       "      <td>2</td>\n",
       "      <td>sort</td>\n",
       "      <td>sanger</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>TRB</td>\n",
       "      <td>CASSYLPGQGDHYSNQPQHF</td>\n",
       "      <td>TRBV13*01</td>\n",
       "      <td>TRBJ1-5*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-B*08</td>\n",
       "      <td>B2M</td>\n",
       "      <td>MHCI</td>\n",
       "      <td>FLKEKGGL</td>\n",
       "      <td>...</td>\n",
       "      <td>HIV-1</td>\n",
       "      <td>PMID:15596521</td>\n",
       "      <td>{\"frequency\": \"\", \"identification\": \"tetramer-...</td>\n",
       "      <td>{\"cell.subset\": \"CD8+\", \"clone.id\": \"\", \"donor...</td>\n",
       "      <td>{\"cdr3\": \"CASSYLPGQGDHYSNQPQHF\", \"cdr3_old\": \"...</td>\n",
       "      <td>2</td>\n",
       "      <td>sort</td>\n",
       "      <td>sanger</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>TRB</td>\n",
       "      <td>CASSFEAGQGFFSNQPQHF</td>\n",
       "      <td>TRBV13*01</td>\n",
       "      <td>TRBJ1-5*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-B*08</td>\n",
       "      <td>B2M</td>\n",
       "      <td>MHCI</td>\n",
       "      <td>FLKEKGGL</td>\n",
       "      <td>...</td>\n",
       "      <td>HIV-1</td>\n",
       "      <td>PMID:15596521</td>\n",
       "      <td>{\"frequency\": \"\", \"identification\": \"tetramer-...</td>\n",
       "      <td>{\"cell.subset\": \"CD8+\", \"clone.id\": \"\", \"donor...</td>\n",
       "      <td>{\"cdr3\": \"CASSFEAGQGFFSNQPQHF\", \"cdr3_old\": \"C...</td>\n",
       "      <td>2</td>\n",
       "      <td>sort</td>\n",
       "      <td>sanger</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2</td>\n",
       "      <td>TRA</td>\n",
       "      <td>CAVPSGAGSYQLTF</td>\n",
       "      <td>TRAV20*01</td>\n",
       "      <td>TRAJ28*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-B*08</td>\n",
       "      <td>B2M</td>\n",
       "      <td>MHCI</td>\n",
       "      <td>FLKEKGGL</td>\n",
       "      <td>...</td>\n",
       "      <td>HIV-1</td>\n",
       "      <td>PMID:15596521</td>\n",
       "      <td>{\"frequency\": \"\", \"identification\": \"tetramer-...</td>\n",
       "      <td>{\"cell.subset\": \"CD8+\", \"clone.id\": \"\", \"donor...</td>\n",
       "      <td>{\"cdr3\": \"CAVPSGAGSYQLTF\", \"cdr3_old\": \"CAVPSG...</td>\n",
       "      <td>2</td>\n",
       "      <td>sort</td>\n",
       "      <td>sanger</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2</td>\n",
       "      <td>TRB</td>\n",
       "      <td>CASSFEPGQGFYSNQPQHF</td>\n",
       "      <td>TRBV13*01</td>\n",
       "      <td>TRBJ1-5*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-B*08</td>\n",
       "      <td>B2M</td>\n",
       "      <td>MHCI</td>\n",
       "      <td>FLKEKGGL</td>\n",
       "      <td>...</td>\n",
       "      <td>HIV-1</td>\n",
       "      <td>PMID:15596521</td>\n",
       "      <td>{\"frequency\": \"\", \"identification\": \"tetramer-...</td>\n",
       "      <td>{\"cell.subset\": \"CD8+\", \"clone.id\": \"\", \"donor...</td>\n",
       "      <td>{\"cdr3\": \"CASSFEPGQGFYSNQPQHF\", \"cdr3_old\": \"C...</td>\n",
       "      <td>2</td>\n",
       "      <td>sort</td>\n",
       "      <td>sanger</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   complex.id gene                  cdr3       v.segm      j.segm  \\\n",
       "0           1  TRA         CIVRAPGRADMRF  TRAV26-1*01   TRAJ43*01   \n",
       "1           1  TRB  CASSYLPGQGDHYSNQPQHF    TRBV13*01  TRBJ1-5*01   \n",
       "2           0  TRB   CASSFEAGQGFFSNQPQHF    TRBV13*01  TRBJ1-5*01   \n",
       "3           2  TRA        CAVPSGAGSYQLTF    TRAV20*01   TRAJ28*01   \n",
       "4           2  TRB   CASSFEPGQGFYSNQPQHF    TRBV13*01  TRBJ1-5*01   \n",
       "\n",
       "       species     mhc.a mhc.b mhc.class antigen.epitope  ... antigen.species  \\\n",
       "0  HomoSapiens  HLA-B*08   B2M      MHCI        FLKEKGGL  ...           HIV-1   \n",
       "1  HomoSapiens  HLA-B*08   B2M      MHCI        FLKEKGGL  ...           HIV-1   \n",
       "2  HomoSapiens  HLA-B*08   B2M      MHCI        FLKEKGGL  ...           HIV-1   \n",
       "3  HomoSapiens  HLA-B*08   B2M      MHCI        FLKEKGGL  ...           HIV-1   \n",
       "4  HomoSapiens  HLA-B*08   B2M      MHCI        FLKEKGGL  ...           HIV-1   \n",
       "\n",
       "    reference.id                                             method  \\\n",
       "0  PMID:15596521  {\"frequency\": \"\", \"identification\": \"tetramer-...   \n",
       "1  PMID:15596521  {\"frequency\": \"\", \"identification\": \"tetramer-...   \n",
       "2  PMID:15596521  {\"frequency\": \"\", \"identification\": \"tetramer-...   \n",
       "3  PMID:15596521  {\"frequency\": \"\", \"identification\": \"tetramer-...   \n",
       "4  PMID:15596521  {\"frequency\": \"\", \"identification\": \"tetramer-...   \n",
       "\n",
       "                                                meta  \\\n",
       "0  {\"cell.subset\": \"CD8+\", \"clone.id\": \"\", \"donor...   \n",
       "1  {\"cell.subset\": \"CD8+\", \"clone.id\": \"\", \"donor...   \n",
       "2  {\"cell.subset\": \"CD8+\", \"clone.id\": \"\", \"donor...   \n",
       "3  {\"cell.subset\": \"CD8+\", \"clone.id\": \"\", \"donor...   \n",
       "4  {\"cell.subset\": \"CD8+\", \"clone.id\": \"\", \"donor...   \n",
       "\n",
       "                                             cdr3fix vdjdb.score  web.method  \\\n",
       "0  {\"cdr3\": \"CIVRAPGRADMRF\", \"cdr3_old\": \"CIVRAPG...           2        sort   \n",
       "1  {\"cdr3\": \"CASSYLPGQGDHYSNQPQHF\", \"cdr3_old\": \"...           2        sort   \n",
       "2  {\"cdr3\": \"CASSFEAGQGFFSNQPQHF\", \"cdr3_old\": \"C...           2        sort   \n",
       "3  {\"cdr3\": \"CAVPSGAGSYQLTF\", \"cdr3_old\": \"CAVPSG...           2        sort   \n",
       "4  {\"cdr3\": \"CASSFEPGQGFYSNQPQHF\", \"cdr3_old\": \"C...           2        sort   \n",
       "\n",
       "  web.method.seq web.cdr3fix.nc web.cdr3fix.unmp  \n",
       "0         sanger             no               no  \n",
       "1         sanger             no               no  \n",
       "2         sanger             no               no  \n",
       "3         sanger             no               no  \n",
       "4         sanger             no               no  \n",
       "\n",
       "[5 rows x 21 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read VDjdb dataset\n",
    "VDJdb = pd.read_csv(\"vdjdb.txt\", sep = \"\\t\")\n",
    "#display(VDJdb)\n",
    "VDJdb.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "5c5aafd8",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cdr3.alpha</th>\n",
       "      <th>v.alpha</th>\n",
       "      <th>j.alpha</th>\n",
       "      <th>cdr3.beta</th>\n",
       "      <th>v.beta</th>\n",
       "      <th>d.beta</th>\n",
       "      <th>j.beta</th>\n",
       "      <th>species</th>\n",
       "      <th>mhc.a</th>\n",
       "      <th>mhc.b</th>\n",
       "      <th>...</th>\n",
       "      <th>meta.replica.id</th>\n",
       "      <th>meta.clone.id</th>\n",
       "      <th>meta.epitope.id</th>\n",
       "      <th>meta.tissue</th>\n",
       "      <th>meta.donor.MHC</th>\n",
       "      <th>meta.donor.MHC.method</th>\n",
       "      <th>meta.structure.id</th>\n",
       "      <th>cdr3fix.alpha</th>\n",
       "      <th>cdr3fix.beta</th>\n",
       "      <th>vdjdb.score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CIVRAPGRADMRF</td>\n",
       "      <td>TRAV26-1*01</td>\n",
       "      <td>TRAJ43*01</td>\n",
       "      <td>CASSYLPGQGDHYSNQPQHF</td>\n",
       "      <td>TRBV13*01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRBJ1-5*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-B*08</td>\n",
       "      <td>B2M</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>HLA-A*02:01,HLA-A*24:02;HLA-B*08:01,HLA-B*5701...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>{\"cdr3\": \"CIVRAPGRADMRF\", \"cdr3_old\": \"CIVRAPG...</td>\n",
       "      <td>{\"cdr3\": \"CASSYLPGQGDHYSNQPQHF\", \"cdr3_old\": \"...</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CASSFEAGQGFFSNQPQHF</td>\n",
       "      <td>TRBV13*01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRBJ1-5*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-B*08</td>\n",
       "      <td>B2M</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>HLA-A*01:01,HLA-A*02:01;HLA-B*08:01,HLA-B*57:0...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>{\"cdr3\": \"CASSFEAGQGFFSNQPQHF\", \"cdr3_old\": \"C...</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CAVPSGAGSYQLTF</td>\n",
       "      <td>TRAV20*01</td>\n",
       "      <td>TRAJ28*01</td>\n",
       "      <td>CASSFEPGQGFYSNQPQHF</td>\n",
       "      <td>TRBV13*01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRBJ1-5*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-B*08</td>\n",
       "      <td>B2M</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>HLA-A*02,HLA-A*24:02;HLA-B*08:01,HLA-B*15;HLA-...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>{\"cdr3\": \"CAVPSGAGSYQLTF\", \"cdr3_old\": \"CAVPSG...</td>\n",
       "      <td>{\"cdr3\": \"CASSFEPGQGFYSNQPQHF\", \"cdr3_old\": \"C...</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CAVKASGSRLT</td>\n",
       "      <td>TRAV2*01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CASSYEPGQVSHYSNQPQHF</td>\n",
       "      <td>TRBV13*01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRBJ1-5*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-B*08</td>\n",
       "      <td>B2M</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>HLA-A*01:01,HLA-A*02:01;HLA-B*08:01,HLA-B*57:0...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>{\"cdr3\": \"CAVKASGSRLT\", \"cdr3_old\": \"CAVKASGSR...</td>\n",
       "      <td>{\"cdr3\": \"CASSYEPGQVSHYSNQPQHF\", \"cdr3_old\": \"...</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CAYRPPGTYKYIF</td>\n",
       "      <td>TRAV38-2/DV8*01</td>\n",
       "      <td>TRAJ40*01</td>\n",
       "      <td>CASSALASLNEQFF</td>\n",
       "      <td>TRBV14*01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRBJ2-1*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-B*08</td>\n",
       "      <td>B2M</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>HLA-A*01:01,HLA-A*02:01;HLA-B*08:01,HLA-B*57:0...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>{\"cdr3\": \"CAYRPPGTYKYIF\", \"cdr3_old\": \"CAYRPPG...</td>\n",
       "      <td>{\"cdr3\": \"CASSALASLNEQFF\", \"cdr3_old\": \"CASSAL...</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60050</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CASSDWGSAETLYF</td>\n",
       "      <td>TRBV13-3*01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRBJ2-3*01</td>\n",
       "      <td>MusMusculus</td>\n",
       "      <td>H-2Db</td>\n",
       "      <td>B2M</td>\n",
       "      <td>...</td>\n",
       "      <td>KL83_CB6F1_#5_21714KL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>H2d/b</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>{\"cdr3\": \"CASSDWGSAETLYF\", \"cdr3_old\": \"CASSDW...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60051</th>\n",
       "      <td>CAARGYNQGKLIF</td>\n",
       "      <td>TRAV14-2*01</td>\n",
       "      <td>TRAJ23*01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MusMusculus</td>\n",
       "      <td>H-2Db</td>\n",
       "      <td>B2M</td>\n",
       "      <td>...</td>\n",
       "      <td>KL136_CB6F1_#5_TRA_271114</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>H2d/b</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>{\"cdr3\": \"CAARGYNQGKLIF\", \"cdr3_old\": \"CAARGYN...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60052</th>\n",
       "      <td>CATSDTNAYKVIF</td>\n",
       "      <td>TRAV8-2*01</td>\n",
       "      <td>TRAJ30*01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MusMusculus</td>\n",
       "      <td>H-2Db</td>\n",
       "      <td>B2M</td>\n",
       "      <td>...</td>\n",
       "      <td>KL136_CB6F1_#5_TRA_271114</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>H2d/b</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>{\"cdr3\": \"CATSDTNAYKVIF\", \"cdr3_old\": \"CATSDTN...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60053</th>\n",
       "      <td>CALSDTGANTGKLTF</td>\n",
       "      <td>TRAV6-5*01</td>\n",
       "      <td>TRAJ52*01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MusMusculus</td>\n",
       "      <td>H-2Db</td>\n",
       "      <td>B2M</td>\n",
       "      <td>...</td>\n",
       "      <td>KL136_CB6F1_#5_TRA_271114</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>H2d/b</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>{\"cdr3\": \"CALSDTGANTGKLTF\", \"cdr3_old\": \"CALSD...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60054</th>\n",
       "      <td>CALRGNNNNRIFF</td>\n",
       "      <td>TRAV9D-4*01</td>\n",
       "      <td>TRAJ31*01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MusMusculus</td>\n",
       "      <td>H-2Db</td>\n",
       "      <td>B2M</td>\n",
       "      <td>...</td>\n",
       "      <td>KL136_CB6F1_#5_TRA_271114</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>H2d/b</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>{\"cdr3\": \"CALRGNNNNRIFF\", \"cdr3_old\": \"CALRGNN...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>60055 rows × 34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            cdr3.alpha          v.alpha    j.alpha             cdr3.beta  \\\n",
       "0        CIVRAPGRADMRF      TRAV26-1*01  TRAJ43*01  CASSYLPGQGDHYSNQPQHF   \n",
       "1                  NaN              NaN        NaN   CASSFEAGQGFFSNQPQHF   \n",
       "2       CAVPSGAGSYQLTF        TRAV20*01  TRAJ28*01   CASSFEPGQGFYSNQPQHF   \n",
       "3          CAVKASGSRLT         TRAV2*01        NaN  CASSYEPGQVSHYSNQPQHF   \n",
       "4        CAYRPPGTYKYIF  TRAV38-2/DV8*01  TRAJ40*01        CASSALASLNEQFF   \n",
       "...                ...              ...        ...                   ...   \n",
       "60050              NaN              NaN        NaN        CASSDWGSAETLYF   \n",
       "60051    CAARGYNQGKLIF      TRAV14-2*01  TRAJ23*01                   NaN   \n",
       "60052    CATSDTNAYKVIF       TRAV8-2*01  TRAJ30*01                   NaN   \n",
       "60053  CALSDTGANTGKLTF       TRAV6-5*01  TRAJ52*01                   NaN   \n",
       "60054    CALRGNNNNRIFF      TRAV9D-4*01  TRAJ31*01                   NaN   \n",
       "\n",
       "            v.beta d.beta      j.beta      species     mhc.a mhc.b  ...  \\\n",
       "0        TRBV13*01    NaN  TRBJ1-5*01  HomoSapiens  HLA-B*08   B2M  ...   \n",
       "1        TRBV13*01    NaN  TRBJ1-5*01  HomoSapiens  HLA-B*08   B2M  ...   \n",
       "2        TRBV13*01    NaN  TRBJ1-5*01  HomoSapiens  HLA-B*08   B2M  ...   \n",
       "3        TRBV13*01    NaN  TRBJ1-5*01  HomoSapiens  HLA-B*08   B2M  ...   \n",
       "4        TRBV14*01    NaN  TRBJ2-1*01  HomoSapiens  HLA-B*08   B2M  ...   \n",
       "...            ...    ...         ...          ...       ...   ...  ...   \n",
       "60050  TRBV13-3*01    NaN  TRBJ2-3*01  MusMusculus     H-2Db   B2M  ...   \n",
       "60051          NaN    NaN         NaN  MusMusculus     H-2Db   B2M  ...   \n",
       "60052          NaN    NaN         NaN  MusMusculus     H-2Db   B2M  ...   \n",
       "60053          NaN    NaN         NaN  MusMusculus     H-2Db   B2M  ...   \n",
       "60054          NaN    NaN         NaN  MusMusculus     H-2Db   B2M  ...   \n",
       "\n",
       "                 meta.replica.id meta.clone.id meta.epitope.id meta.tissue  \\\n",
       "0                            NaN           NaN             NaN        PBMC   \n",
       "1                            NaN           NaN             NaN        PBMC   \n",
       "2                            NaN           NaN             NaN        PBMC   \n",
       "3                            NaN           NaN             NaN        PBMC   \n",
       "4                            NaN           NaN             NaN        PBMC   \n",
       "...                          ...           ...             ...         ...   \n",
       "60050      KL83_CB6F1_#5_21714KL           NaN             NaN         NaN   \n",
       "60051  KL136_CB6F1_#5_TRA_271114           NaN             NaN         NaN   \n",
       "60052  KL136_CB6F1_#5_TRA_271114           NaN             NaN         NaN   \n",
       "60053  KL136_CB6F1_#5_TRA_271114           NaN             NaN         NaN   \n",
       "60054  KL136_CB6F1_#5_TRA_271114           NaN             NaN         NaN   \n",
       "\n",
       "                                          meta.donor.MHC  \\\n",
       "0      HLA-A*02:01,HLA-A*24:02;HLA-B*08:01,HLA-B*5701...   \n",
       "1      HLA-A*01:01,HLA-A*02:01;HLA-B*08:01,HLA-B*57:0...   \n",
       "2      HLA-A*02,HLA-A*24:02;HLA-B*08:01,HLA-B*15;HLA-...   \n",
       "3      HLA-A*01:01,HLA-A*02:01;HLA-B*08:01,HLA-B*57:0...   \n",
       "4      HLA-A*01:01,HLA-A*02:01;HLA-B*08:01,HLA-B*57:0...   \n",
       "...                                                  ...   \n",
       "60050                                              H2d/b   \n",
       "60051                                              H2d/b   \n",
       "60052                                              H2d/b   \n",
       "60053                                              H2d/b   \n",
       "60054                                              H2d/b   \n",
       "\n",
       "      meta.donor.MHC.method meta.structure.id  \\\n",
       "0                       NaN               NaN   \n",
       "1                       NaN               NaN   \n",
       "2                       NaN               NaN   \n",
       "3                       NaN               NaN   \n",
       "4                       NaN               NaN   \n",
       "...                     ...               ...   \n",
       "60050                   NaN               NaN   \n",
       "60051                   NaN               NaN   \n",
       "60052                   NaN               NaN   \n",
       "60053                   NaN               NaN   \n",
       "60054                   NaN               NaN   \n",
       "\n",
       "                                           cdr3fix.alpha  \\\n",
       "0      {\"cdr3\": \"CIVRAPGRADMRF\", \"cdr3_old\": \"CIVRAPG...   \n",
       "1                                                    NaN   \n",
       "2      {\"cdr3\": \"CAVPSGAGSYQLTF\", \"cdr3_old\": \"CAVPSG...   \n",
       "3      {\"cdr3\": \"CAVKASGSRLT\", \"cdr3_old\": \"CAVKASGSR...   \n",
       "4      {\"cdr3\": \"CAYRPPGTYKYIF\", \"cdr3_old\": \"CAYRPPG...   \n",
       "...                                                  ...   \n",
       "60050                                                NaN   \n",
       "60051  {\"cdr3\": \"CAARGYNQGKLIF\", \"cdr3_old\": \"CAARGYN...   \n",
       "60052  {\"cdr3\": \"CATSDTNAYKVIF\", \"cdr3_old\": \"CATSDTN...   \n",
       "60053  {\"cdr3\": \"CALSDTGANTGKLTF\", \"cdr3_old\": \"CALSD...   \n",
       "60054  {\"cdr3\": \"CALRGNNNNRIFF\", \"cdr3_old\": \"CALRGNN...   \n",
       "\n",
       "                                            cdr3fix.beta vdjdb.score  \n",
       "0      {\"cdr3\": \"CASSYLPGQGDHYSNQPQHF\", \"cdr3_old\": \"...           2  \n",
       "1      {\"cdr3\": \"CASSFEAGQGFFSNQPQHF\", \"cdr3_old\": \"C...           2  \n",
       "2      {\"cdr3\": \"CASSFEPGQGFYSNQPQHF\", \"cdr3_old\": \"C...           2  \n",
       "3      {\"cdr3\": \"CASSYEPGQVSHYSNQPQHF\", \"cdr3_old\": \"...           2  \n",
       "4      {\"cdr3\": \"CASSALASLNEQFF\", \"cdr3_old\": \"CASSAL...           2  \n",
       "...                                                  ...         ...  \n",
       "60050  {\"cdr3\": \"CASSDWGSAETLYF\", \"cdr3_old\": \"CASSDW...           0  \n",
       "60051                                                NaN           1  \n",
       "60052                                                NaN           1  \n",
       "60053                                                NaN           0  \n",
       "60054                                                NaN           0  \n",
       "\n",
       "[60055 rows x 34 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# read VDJdb full dataset\n",
    "VDJdb_full = pd.read_csv(\"vdjdb_full.txt\", sep = \"\\t\", low_memory=False)\n",
    "display(VDJdb_full)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "af85689e",
   "metadata": {},
   "source": [
    "Based on my understanding of the original data and its background knowledge. Its useful data are listed as (cdr3.alpha, v.alpha, j.alpha, cdr3.beta, v.beta, d.beta, j.beta, species).  \n",
    "Our second step is to derive the similarity matrix from the useful sequence data by means of a library of functions such as TCRDist. For documentation on the TCRDist functions, please click on https://tcrdist3.readthedocs.io/en/latest/welcome.html.  \n",
    "\n",
    "Here, I am not sure if the v,j segment needs to be converted to a nucleotide sequence when calculating the similarity matrix for the TCR. If the v, j segment needs to be converted to a nucleotide sequence, please click on https://www.ncbi.nlm.nih.gov/pmc/articles/PMC9262623/. For the relevant Github code, and associated conversion data, click: https://github.com/JamieHeather/stitchr.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "dfaa1893",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>cdr3.alpha</th>\n",
       "      <th>v.alpha</th>\n",
       "      <th>j.alpha</th>\n",
       "      <th>cdr3.beta</th>\n",
       "      <th>v.beta</th>\n",
       "      <th>j.beta</th>\n",
       "      <th>mhc.a</th>\n",
       "      <th>mhc.b</th>\n",
       "      <th>vdjdb.score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>CIVRAPGRADMRF</td>\n",
       "      <td>TRAV26-1*01</td>\n",
       "      <td>TRAJ43*01</td>\n",
       "      <td>CASSYLPGQGDHYSNQPQHF</td>\n",
       "      <td>TRBV13*01</td>\n",
       "      <td>TRBJ1-5*01</td>\n",
       "      <td>HLA-B*08</td>\n",
       "      <td>B2M</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CASSFEAGQGFFSNQPQHF</td>\n",
       "      <td>TRBV13*01</td>\n",
       "      <td>TRBJ1-5*01</td>\n",
       "      <td>HLA-B*08</td>\n",
       "      <td>B2M</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>CAVPSGAGSYQLTF</td>\n",
       "      <td>TRAV20*01</td>\n",
       "      <td>TRAJ28*01</td>\n",
       "      <td>CASSFEPGQGFYSNQPQHF</td>\n",
       "      <td>TRBV13*01</td>\n",
       "      <td>TRBJ1-5*01</td>\n",
       "      <td>HLA-B*08</td>\n",
       "      <td>B2M</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>CAVKASGSRLT</td>\n",
       "      <td>TRAV2*01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CASSYEPGQVSHYSNQPQHF</td>\n",
       "      <td>TRBV13*01</td>\n",
       "      <td>TRBJ1-5*01</td>\n",
       "      <td>HLA-B*08</td>\n",
       "      <td>B2M</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>CAYRPPGTYKYIF</td>\n",
       "      <td>TRAV38-2/DV8*01</td>\n",
       "      <td>TRAJ40*01</td>\n",
       "      <td>CASSALASLNEQFF</td>\n",
       "      <td>TRBV14*01</td>\n",
       "      <td>TRBJ2-1*01</td>\n",
       "      <td>HLA-B*08</td>\n",
       "      <td>B2M</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60050</th>\n",
       "      <td>MusMusculus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CASSDWGSAETLYF</td>\n",
       "      <td>TRBV13-3*01</td>\n",
       "      <td>TRBJ2-3*01</td>\n",
       "      <td>H-2Db</td>\n",
       "      <td>B2M</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60051</th>\n",
       "      <td>MusMusculus</td>\n",
       "      <td>CAARGYNQGKLIF</td>\n",
       "      <td>TRAV14-2*01</td>\n",
       "      <td>TRAJ23*01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>H-2Db</td>\n",
       "      <td>B2M</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60052</th>\n",
       "      <td>MusMusculus</td>\n",
       "      <td>CATSDTNAYKVIF</td>\n",
       "      <td>TRAV8-2*01</td>\n",
       "      <td>TRAJ30*01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>H-2Db</td>\n",
       "      <td>B2M</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60053</th>\n",
       "      <td>MusMusculus</td>\n",
       "      <td>CALSDTGANTGKLTF</td>\n",
       "      <td>TRAV6-5*01</td>\n",
       "      <td>TRAJ52*01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>H-2Db</td>\n",
       "      <td>B2M</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60054</th>\n",
       "      <td>MusMusculus</td>\n",
       "      <td>CALRGNNNNRIFF</td>\n",
       "      <td>TRAV9D-4*01</td>\n",
       "      <td>TRAJ31*01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>H-2Db</td>\n",
       "      <td>B2M</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>60055 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           species       cdr3.alpha          v.alpha    j.alpha  \\\n",
       "0      HomoSapiens    CIVRAPGRADMRF      TRAV26-1*01  TRAJ43*01   \n",
       "1      HomoSapiens              NaN              NaN        NaN   \n",
       "2      HomoSapiens   CAVPSGAGSYQLTF        TRAV20*01  TRAJ28*01   \n",
       "3      HomoSapiens      CAVKASGSRLT         TRAV2*01        NaN   \n",
       "4      HomoSapiens    CAYRPPGTYKYIF  TRAV38-2/DV8*01  TRAJ40*01   \n",
       "...            ...              ...              ...        ...   \n",
       "60050  MusMusculus              NaN              NaN        NaN   \n",
       "60051  MusMusculus    CAARGYNQGKLIF      TRAV14-2*01  TRAJ23*01   \n",
       "60052  MusMusculus    CATSDTNAYKVIF       TRAV8-2*01  TRAJ30*01   \n",
       "60053  MusMusculus  CALSDTGANTGKLTF       TRAV6-5*01  TRAJ52*01   \n",
       "60054  MusMusculus    CALRGNNNNRIFF      TRAV9D-4*01  TRAJ31*01   \n",
       "\n",
       "                  cdr3.beta       v.beta      j.beta     mhc.a mhc.b  \\\n",
       "0      CASSYLPGQGDHYSNQPQHF    TRBV13*01  TRBJ1-5*01  HLA-B*08   B2M   \n",
       "1       CASSFEAGQGFFSNQPQHF    TRBV13*01  TRBJ1-5*01  HLA-B*08   B2M   \n",
       "2       CASSFEPGQGFYSNQPQHF    TRBV13*01  TRBJ1-5*01  HLA-B*08   B2M   \n",
       "3      CASSYEPGQVSHYSNQPQHF    TRBV13*01  TRBJ1-5*01  HLA-B*08   B2M   \n",
       "4            CASSALASLNEQFF    TRBV14*01  TRBJ2-1*01  HLA-B*08   B2M   \n",
       "...                     ...          ...         ...       ...   ...   \n",
       "60050        CASSDWGSAETLYF  TRBV13-3*01  TRBJ2-3*01     H-2Db   B2M   \n",
       "60051                   NaN          NaN         NaN     H-2Db   B2M   \n",
       "60052                   NaN          NaN         NaN     H-2Db   B2M   \n",
       "60053                   NaN          NaN         NaN     H-2Db   B2M   \n",
       "60054                   NaN          NaN         NaN     H-2Db   B2M   \n",
       "\n",
       "       vdjdb.score  \n",
       "0                2  \n",
       "1                2  \n",
       "2                2  \n",
       "3                2  \n",
       "4                2  \n",
       "...            ...  \n",
       "60050            0  \n",
       "60051            1  \n",
       "60052            1  \n",
       "60053            0  \n",
       "60054            0  \n",
       "\n",
       "[60055 rows x 10 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>cdr3_a_aa</th>\n",
       "      <th>v_a_gene</th>\n",
       "      <th>j_a_gene</th>\n",
       "      <th>cdr3_b_aa</th>\n",
       "      <th>v_b_gene</th>\n",
       "      <th>j_b_gene</th>\n",
       "      <th>mhc.a</th>\n",
       "      <th>mhc.b</th>\n",
       "      <th>vdjdb.score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>CIVRAPGRADMRF</td>\n",
       "      <td>TRAV26-1*01</td>\n",
       "      <td>TRAJ43*01</td>\n",
       "      <td>CASSYLPGQGDHYSNQPQHF</td>\n",
       "      <td>TRBV13*01</td>\n",
       "      <td>TRBJ1-5*01</td>\n",
       "      <td>HLA-B*08</td>\n",
       "      <td>B2M</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CASSFEAGQGFFSNQPQHF</td>\n",
       "      <td>TRBV13*01</td>\n",
       "      <td>TRBJ1-5*01</td>\n",
       "      <td>HLA-B*08</td>\n",
       "      <td>B2M</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>CAVPSGAGSYQLTF</td>\n",
       "      <td>TRAV20*01</td>\n",
       "      <td>TRAJ28*01</td>\n",
       "      <td>CASSFEPGQGFYSNQPQHF</td>\n",
       "      <td>TRBV13*01</td>\n",
       "      <td>TRBJ1-5*01</td>\n",
       "      <td>HLA-B*08</td>\n",
       "      <td>B2M</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>CAVKASGSRLT</td>\n",
       "      <td>TRAV2*01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CASSYEPGQVSHYSNQPQHF</td>\n",
       "      <td>TRBV13*01</td>\n",
       "      <td>TRBJ1-5*01</td>\n",
       "      <td>HLA-B*08</td>\n",
       "      <td>B2M</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>CAYRPPGTYKYIF</td>\n",
       "      <td>TRAV38-2/DV8*01</td>\n",
       "      <td>TRAJ40*01</td>\n",
       "      <td>CASSALASLNEQFF</td>\n",
       "      <td>TRBV14*01</td>\n",
       "      <td>TRBJ2-1*01</td>\n",
       "      <td>HLA-B*08</td>\n",
       "      <td>B2M</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>495</th>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CASSYFGVNSPLHF</td>\n",
       "      <td>TRBV6-1*01</td>\n",
       "      <td>TRBJ1-6*01</td>\n",
       "      <td>HLA-A*02:01</td>\n",
       "      <td>B2M</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>496</th>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CATSALAGQGRDEQFF</td>\n",
       "      <td>TRBV24-1*01</td>\n",
       "      <td>TRBJ2-1*01</td>\n",
       "      <td>HLA-A*02:01</td>\n",
       "      <td>B2M</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>497</th>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CASSRLAGTDTQYF</td>\n",
       "      <td>TRBV12-3*01</td>\n",
       "      <td>TRBJ2-3*01</td>\n",
       "      <td>HLA-A*02:01</td>\n",
       "      <td>B2M</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>498</th>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CASSFPGYGLNTEAFF</td>\n",
       "      <td>TRBV6-1*01</td>\n",
       "      <td>TRBJ1-1*01</td>\n",
       "      <td>HLA-A*02:01</td>\n",
       "      <td>B2M</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>499</th>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CASSVLAGGLDTQYF</td>\n",
       "      <td>TRBV10-2*01</td>\n",
       "      <td>TRBJ2-3*01</td>\n",
       "      <td>HLA-A*02:01</td>\n",
       "      <td>B2M</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>500 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         species       cdr3_a_aa         v_a_gene   j_a_gene  \\\n",
       "0    HomoSapiens   CIVRAPGRADMRF      TRAV26-1*01  TRAJ43*01   \n",
       "1    HomoSapiens             NaN              NaN        NaN   \n",
       "2    HomoSapiens  CAVPSGAGSYQLTF        TRAV20*01  TRAJ28*01   \n",
       "3    HomoSapiens     CAVKASGSRLT         TRAV2*01        NaN   \n",
       "4    HomoSapiens   CAYRPPGTYKYIF  TRAV38-2/DV8*01  TRAJ40*01   \n",
       "..           ...             ...              ...        ...   \n",
       "495  HomoSapiens             NaN              NaN        NaN   \n",
       "496  HomoSapiens             NaN              NaN        NaN   \n",
       "497  HomoSapiens             NaN              NaN        NaN   \n",
       "498  HomoSapiens             NaN              NaN        NaN   \n",
       "499  HomoSapiens             NaN              NaN        NaN   \n",
       "\n",
       "                cdr3_b_aa     v_b_gene    j_b_gene        mhc.a mhc.b  \\\n",
       "0    CASSYLPGQGDHYSNQPQHF    TRBV13*01  TRBJ1-5*01     HLA-B*08   B2M   \n",
       "1     CASSFEAGQGFFSNQPQHF    TRBV13*01  TRBJ1-5*01     HLA-B*08   B2M   \n",
       "2     CASSFEPGQGFYSNQPQHF    TRBV13*01  TRBJ1-5*01     HLA-B*08   B2M   \n",
       "3    CASSYEPGQVSHYSNQPQHF    TRBV13*01  TRBJ1-5*01     HLA-B*08   B2M   \n",
       "4          CASSALASLNEQFF    TRBV14*01  TRBJ2-1*01     HLA-B*08   B2M   \n",
       "..                    ...          ...         ...          ...   ...   \n",
       "495        CASSYFGVNSPLHF   TRBV6-1*01  TRBJ1-6*01  HLA-A*02:01   B2M   \n",
       "496      CATSALAGQGRDEQFF  TRBV24-1*01  TRBJ2-1*01  HLA-A*02:01   B2M   \n",
       "497        CASSRLAGTDTQYF  TRBV12-3*01  TRBJ2-3*01  HLA-A*02:01   B2M   \n",
       "498      CASSFPGYGLNTEAFF   TRBV6-1*01  TRBJ1-1*01  HLA-A*02:01   B2M   \n",
       "499       CASSVLAGGLDTQYF  TRBV10-2*01  TRBJ2-3*01  HLA-A*02:01   B2M   \n",
       "\n",
       "     vdjdb.score  \n",
       "0              2  \n",
       "1              2  \n",
       "2              2  \n",
       "3              2  \n",
       "4              2  \n",
       "..           ...  \n",
       "495            0  \n",
       "496            0  \n",
       "497            0  \n",
       "498            0  \n",
       "499            0  \n",
       "\n",
       "[500 rows x 10 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "vdj_useful_db = VDJdb_full[[\"species\", \"cdr3.alpha\", \n",
    "                            \"v.alpha\", \"j.alpha\", \"cdr3.beta\", \n",
    "                            \"v.beta\", \"j.beta\", \"mhc.a\", \"mhc.b\", \"vdjdb.score\"]]\n",
    "display(vdj_useful_db)\n",
    "\n",
    "vdj_db_fit_tcrdist = vdj_useful_db.rename(columns = {'cdr3.alpha': 'cdr3_a_aa', 'v.alpha': 'v_a_gene', 'j.alpha':'j_a_gene', \n",
    "                                          'cdr3.beta': 'cdr3_b_aa', 'v.beta':'v_b_gene', 'j.beta':'j_b_gene'})\n",
    "\n",
    "vdj_db_fit_tcrdist_first_500 = vdj_db_fit_tcrdist.head(500)\n",
    "display(vdj_db_fit_tcrdist_first_500)\n",
    "\n",
    "full_db = VDJdb_full"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "79f073f1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>complex.id</th>\n",
       "      <th>gene</th>\n",
       "      <th>cdr3</th>\n",
       "      <th>v.segm</th>\n",
       "      <th>j.segm</th>\n",
       "      <th>species</th>\n",
       "      <th>mhc.a</th>\n",
       "      <th>mhc.b</th>\n",
       "      <th>mhc.class</th>\n",
       "      <th>antigen.epitope</th>\n",
       "      <th>...</th>\n",
       "      <th>antigen.species</th>\n",
       "      <th>reference.id</th>\n",
       "      <th>method</th>\n",
       "      <th>meta</th>\n",
       "      <th>cdr3fix</th>\n",
       "      <th>vdjdb.score</th>\n",
       "      <th>web.method</th>\n",
       "      <th>web.method.seq</th>\n",
       "      <th>web.cdr3fix.nc</th>\n",
       "      <th>web.cdr3fix.unmp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>TRA</td>\n",
       "      <td>CIVRAPGRADMRF</td>\n",
       "      <td>TRAV26-1*01</td>\n",
       "      <td>TRAJ43*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-B*08</td>\n",
       "      <td>B2M</td>\n",
       "      <td>MHCI</td>\n",
       "      <td>FLKEKGGL</td>\n",
       "      <td>...</td>\n",
       "      <td>HIV-1</td>\n",
       "      <td>PMID:15596521</td>\n",
       "      <td>{\"frequency\": \"\", \"identification\": \"tetramer-...</td>\n",
       "      <td>{\"cell.subset\": \"CD8+\", \"clone.id\": \"\", \"donor...</td>\n",
       "      <td>{\"cdr3\": \"CIVRAPGRADMRF\", \"cdr3_old\": \"CIVRAPG...</td>\n",
       "      <td>2</td>\n",
       "      <td>sort</td>\n",
       "      <td>sanger</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>TRB</td>\n",
       "      <td>CASSYLPGQGDHYSNQPQHF</td>\n",
       "      <td>TRBV13*01</td>\n",
       "      <td>TRBJ1-5*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-B*08</td>\n",
       "      <td>B2M</td>\n",
       "      <td>MHCI</td>\n",
       "      <td>FLKEKGGL</td>\n",
       "      <td>...</td>\n",
       "      <td>HIV-1</td>\n",
       "      <td>PMID:15596521</td>\n",
       "      <td>{\"frequency\": \"\", \"identification\": \"tetramer-...</td>\n",
       "      <td>{\"cell.subset\": \"CD8+\", \"clone.id\": \"\", \"donor...</td>\n",
       "      <td>{\"cdr3\": \"CASSYLPGQGDHYSNQPQHF\", \"cdr3_old\": \"...</td>\n",
       "      <td>2</td>\n",
       "      <td>sort</td>\n",
       "      <td>sanger</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>TRB</td>\n",
       "      <td>CASSFEAGQGFFSNQPQHF</td>\n",
       "      <td>TRBV13*01</td>\n",
       "      <td>TRBJ1-5*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-B*08</td>\n",
       "      <td>B2M</td>\n",
       "      <td>MHCI</td>\n",
       "      <td>FLKEKGGL</td>\n",
       "      <td>...</td>\n",
       "      <td>HIV-1</td>\n",
       "      <td>PMID:15596521</td>\n",
       "      <td>{\"frequency\": \"\", \"identification\": \"tetramer-...</td>\n",
       "      <td>{\"cell.subset\": \"CD8+\", \"clone.id\": \"\", \"donor...</td>\n",
       "      <td>{\"cdr3\": \"CASSFEAGQGFFSNQPQHF\", \"cdr3_old\": \"C...</td>\n",
       "      <td>2</td>\n",
       "      <td>sort</td>\n",
       "      <td>sanger</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2</td>\n",
       "      <td>TRA</td>\n",
       "      <td>CAVPSGAGSYQLTF</td>\n",
       "      <td>TRAV20*01</td>\n",
       "      <td>TRAJ28*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-B*08</td>\n",
       "      <td>B2M</td>\n",
       "      <td>MHCI</td>\n",
       "      <td>FLKEKGGL</td>\n",
       "      <td>...</td>\n",
       "      <td>HIV-1</td>\n",
       "      <td>PMID:15596521</td>\n",
       "      <td>{\"frequency\": \"\", \"identification\": \"tetramer-...</td>\n",
       "      <td>{\"cell.subset\": \"CD8+\", \"clone.id\": \"\", \"donor...</td>\n",
       "      <td>{\"cdr3\": \"CAVPSGAGSYQLTF\", \"cdr3_old\": \"CAVPSG...</td>\n",
       "      <td>2</td>\n",
       "      <td>sort</td>\n",
       "      <td>sanger</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2</td>\n",
       "      <td>TRB</td>\n",
       "      <td>CASSFEPGQGFYSNQPQHF</td>\n",
       "      <td>TRBV13*01</td>\n",
       "      <td>TRBJ1-5*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-B*08</td>\n",
       "      <td>B2M</td>\n",
       "      <td>MHCI</td>\n",
       "      <td>FLKEKGGL</td>\n",
       "      <td>...</td>\n",
       "      <td>HIV-1</td>\n",
       "      <td>PMID:15596521</td>\n",
       "      <td>{\"frequency\": \"\", \"identification\": \"tetramer-...</td>\n",
       "      <td>{\"cell.subset\": \"CD8+\", \"clone.id\": \"\", \"donor...</td>\n",
       "      <td>{\"cdr3\": \"CASSFEPGQGFYSNQPQHF\", \"cdr3_old\": \"C...</td>\n",
       "      <td>2</td>\n",
       "      <td>sort</td>\n",
       "      <td>sanger</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89316</th>\n",
       "      <td>0</td>\n",
       "      <td>TRB</td>\n",
       "      <td>CASSDWGSAETLYF</td>\n",
       "      <td>TRBV13-3*01</td>\n",
       "      <td>TRBJ2-3*01</td>\n",
       "      <td>MusMusculus</td>\n",
       "      <td>H-2Db</td>\n",
       "      <td>B2M</td>\n",
       "      <td>MHCI</td>\n",
       "      <td>SQLLNAKYL</td>\n",
       "      <td>...</td>\n",
       "      <td>PlasmodiumBerghei</td>\n",
       "      <td>PMID:29150238</td>\n",
       "      <td>{\"frequency\": \"1//75\", \"identification\": \"tetr...</td>\n",
       "      <td>{\"cell.subset\": \"CD8\", \"clone.id\": \"\", \"donor....</td>\n",
       "      <td>{\"cdr3\": \"CASSDWGSAETLYF\", \"cdr3_old\": \"CASSDW...</td>\n",
       "      <td>0</td>\n",
       "      <td>sort</td>\n",
       "      <td>sanger</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89317</th>\n",
       "      <td>0</td>\n",
       "      <td>TRA</td>\n",
       "      <td>CAARGYNQGKLIF</td>\n",
       "      <td>TRAV14-2*01</td>\n",
       "      <td>TRAJ23*01</td>\n",
       "      <td>MusMusculus</td>\n",
       "      <td>H-2Db</td>\n",
       "      <td>B2M</td>\n",
       "      <td>MHCI</td>\n",
       "      <td>SQLLNAKYL</td>\n",
       "      <td>...</td>\n",
       "      <td>PlasmodiumBerghei</td>\n",
       "      <td>PMID:29150238</td>\n",
       "      <td>{\"frequency\": \"48//86\", \"identification\": \"tet...</td>\n",
       "      <td>{\"cell.subset\": \"CD8\", \"clone.id\": \"\", \"donor....</td>\n",
       "      <td>{\"cdr3\": \"CAARGYNQGKLIF\", \"cdr3_old\": \"CAARGYN...</td>\n",
       "      <td>1</td>\n",
       "      <td>sort</td>\n",
       "      <td>sanger</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89318</th>\n",
       "      <td>0</td>\n",
       "      <td>TRA</td>\n",
       "      <td>CATSDTNAYKVIF</td>\n",
       "      <td>TRAV8-2*01</td>\n",
       "      <td>TRAJ30*01</td>\n",
       "      <td>MusMusculus</td>\n",
       "      <td>H-2Db</td>\n",
       "      <td>B2M</td>\n",
       "      <td>MHCI</td>\n",
       "      <td>SQLLNAKYL</td>\n",
       "      <td>...</td>\n",
       "      <td>PlasmodiumBerghei</td>\n",
       "      <td>PMID:29150238</td>\n",
       "      <td>{\"frequency\": \"36//86\", \"identification\": \"tet...</td>\n",
       "      <td>{\"cell.subset\": \"CD8\", \"clone.id\": \"\", \"donor....</td>\n",
       "      <td>{\"cdr3\": \"CATSDTNAYKVIF\", \"cdr3_old\": \"CATSDTN...</td>\n",
       "      <td>1</td>\n",
       "      <td>sort</td>\n",
       "      <td>sanger</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89319</th>\n",
       "      <td>0</td>\n",
       "      <td>TRA</td>\n",
       "      <td>CALSDTGANTGKLTF</td>\n",
       "      <td>TRAV6-5*01</td>\n",
       "      <td>TRAJ52*01</td>\n",
       "      <td>MusMusculus</td>\n",
       "      <td>H-2Db</td>\n",
       "      <td>B2M</td>\n",
       "      <td>MHCI</td>\n",
       "      <td>SQLLNAKYL</td>\n",
       "      <td>...</td>\n",
       "      <td>PlasmodiumBerghei</td>\n",
       "      <td>PMID:29150238</td>\n",
       "      <td>{\"frequency\": \"1//86\", \"identification\": \"tetr...</td>\n",
       "      <td>{\"cell.subset\": \"CD8\", \"clone.id\": \"\", \"donor....</td>\n",
       "      <td>{\"cdr3\": \"CALSDTGANTGKLTF\", \"cdr3_old\": \"CALSD...</td>\n",
       "      <td>0</td>\n",
       "      <td>sort</td>\n",
       "      <td>sanger</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89320</th>\n",
       "      <td>0</td>\n",
       "      <td>TRA</td>\n",
       "      <td>CALRGNNNNRIFF</td>\n",
       "      <td>TRAV9D-4*01</td>\n",
       "      <td>TRAJ31*01</td>\n",
       "      <td>MusMusculus</td>\n",
       "      <td>H-2Db</td>\n",
       "      <td>B2M</td>\n",
       "      <td>MHCI</td>\n",
       "      <td>SQLLNAKYL</td>\n",
       "      <td>...</td>\n",
       "      <td>PlasmodiumBerghei</td>\n",
       "      <td>PMID:29150238</td>\n",
       "      <td>{\"frequency\": \"1//86\", \"identification\": \"tetr...</td>\n",
       "      <td>{\"cell.subset\": \"CD8\", \"clone.id\": \"\", \"donor....</td>\n",
       "      <td>{\"cdr3\": \"CALRGNNNNRIFF\", \"cdr3_old\": \"CALRGNN...</td>\n",
       "      <td>0</td>\n",
       "      <td>sort</td>\n",
       "      <td>sanger</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>89321 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       complex.id gene                  cdr3       v.segm      j.segm  \\\n",
       "0               1  TRA         CIVRAPGRADMRF  TRAV26-1*01   TRAJ43*01   \n",
       "1               1  TRB  CASSYLPGQGDHYSNQPQHF    TRBV13*01  TRBJ1-5*01   \n",
       "2               0  TRB   CASSFEAGQGFFSNQPQHF    TRBV13*01  TRBJ1-5*01   \n",
       "3               2  TRA        CAVPSGAGSYQLTF    TRAV20*01   TRAJ28*01   \n",
       "4               2  TRB   CASSFEPGQGFYSNQPQHF    TRBV13*01  TRBJ1-5*01   \n",
       "...           ...  ...                   ...          ...         ...   \n",
       "89316           0  TRB        CASSDWGSAETLYF  TRBV13-3*01  TRBJ2-3*01   \n",
       "89317           0  TRA         CAARGYNQGKLIF  TRAV14-2*01   TRAJ23*01   \n",
       "89318           0  TRA         CATSDTNAYKVIF   TRAV8-2*01   TRAJ30*01   \n",
       "89319           0  TRA       CALSDTGANTGKLTF   TRAV6-5*01   TRAJ52*01   \n",
       "89320           0  TRA         CALRGNNNNRIFF  TRAV9D-4*01   TRAJ31*01   \n",
       "\n",
       "           species     mhc.a mhc.b mhc.class antigen.epitope  ...  \\\n",
       "0      HomoSapiens  HLA-B*08   B2M      MHCI        FLKEKGGL  ...   \n",
       "1      HomoSapiens  HLA-B*08   B2M      MHCI        FLKEKGGL  ...   \n",
       "2      HomoSapiens  HLA-B*08   B2M      MHCI        FLKEKGGL  ...   \n",
       "3      HomoSapiens  HLA-B*08   B2M      MHCI        FLKEKGGL  ...   \n",
       "4      HomoSapiens  HLA-B*08   B2M      MHCI        FLKEKGGL  ...   \n",
       "...            ...       ...   ...       ...             ...  ...   \n",
       "89316  MusMusculus     H-2Db   B2M      MHCI       SQLLNAKYL  ...   \n",
       "89317  MusMusculus     H-2Db   B2M      MHCI       SQLLNAKYL  ...   \n",
       "89318  MusMusculus     H-2Db   B2M      MHCI       SQLLNAKYL  ...   \n",
       "89319  MusMusculus     H-2Db   B2M      MHCI       SQLLNAKYL  ...   \n",
       "89320  MusMusculus     H-2Db   B2M      MHCI       SQLLNAKYL  ...   \n",
       "\n",
       "         antigen.species   reference.id  \\\n",
       "0                  HIV-1  PMID:15596521   \n",
       "1                  HIV-1  PMID:15596521   \n",
       "2                  HIV-1  PMID:15596521   \n",
       "3                  HIV-1  PMID:15596521   \n",
       "4                  HIV-1  PMID:15596521   \n",
       "...                  ...            ...   \n",
       "89316  PlasmodiumBerghei  PMID:29150238   \n",
       "89317  PlasmodiumBerghei  PMID:29150238   \n",
       "89318  PlasmodiumBerghei  PMID:29150238   \n",
       "89319  PlasmodiumBerghei  PMID:29150238   \n",
       "89320  PlasmodiumBerghei  PMID:29150238   \n",
       "\n",
       "                                                  method  \\\n",
       "0      {\"frequency\": \"\", \"identification\": \"tetramer-...   \n",
       "1      {\"frequency\": \"\", \"identification\": \"tetramer-...   \n",
       "2      {\"frequency\": \"\", \"identification\": \"tetramer-...   \n",
       "3      {\"frequency\": \"\", \"identification\": \"tetramer-...   \n",
       "4      {\"frequency\": \"\", \"identification\": \"tetramer-...   \n",
       "...                                                  ...   \n",
       "89316  {\"frequency\": \"1//75\", \"identification\": \"tetr...   \n",
       "89317  {\"frequency\": \"48//86\", \"identification\": \"tet...   \n",
       "89318  {\"frequency\": \"36//86\", \"identification\": \"tet...   \n",
       "89319  {\"frequency\": \"1//86\", \"identification\": \"tetr...   \n",
       "89320  {\"frequency\": \"1//86\", \"identification\": \"tetr...   \n",
       "\n",
       "                                                    meta  \\\n",
       "0      {\"cell.subset\": \"CD8+\", \"clone.id\": \"\", \"donor...   \n",
       "1      {\"cell.subset\": \"CD8+\", \"clone.id\": \"\", \"donor...   \n",
       "2      {\"cell.subset\": \"CD8+\", \"clone.id\": \"\", \"donor...   \n",
       "3      {\"cell.subset\": \"CD8+\", \"clone.id\": \"\", \"donor...   \n",
       "4      {\"cell.subset\": \"CD8+\", \"clone.id\": \"\", \"donor...   \n",
       "...                                                  ...   \n",
       "89316  {\"cell.subset\": \"CD8\", \"clone.id\": \"\", \"donor....   \n",
       "89317  {\"cell.subset\": \"CD8\", \"clone.id\": \"\", \"donor....   \n",
       "89318  {\"cell.subset\": \"CD8\", \"clone.id\": \"\", \"donor....   \n",
       "89319  {\"cell.subset\": \"CD8\", \"clone.id\": \"\", \"donor....   \n",
       "89320  {\"cell.subset\": \"CD8\", \"clone.id\": \"\", \"donor....   \n",
       "\n",
       "                                                 cdr3fix vdjdb.score  \\\n",
       "0      {\"cdr3\": \"CIVRAPGRADMRF\", \"cdr3_old\": \"CIVRAPG...           2   \n",
       "1      {\"cdr3\": \"CASSYLPGQGDHYSNQPQHF\", \"cdr3_old\": \"...           2   \n",
       "2      {\"cdr3\": \"CASSFEAGQGFFSNQPQHF\", \"cdr3_old\": \"C...           2   \n",
       "3      {\"cdr3\": \"CAVPSGAGSYQLTF\", \"cdr3_old\": \"CAVPSG...           2   \n",
       "4      {\"cdr3\": \"CASSFEPGQGFYSNQPQHF\", \"cdr3_old\": \"C...           2   \n",
       "...                                                  ...         ...   \n",
       "89316  {\"cdr3\": \"CASSDWGSAETLYF\", \"cdr3_old\": \"CASSDW...           0   \n",
       "89317  {\"cdr3\": \"CAARGYNQGKLIF\", \"cdr3_old\": \"CAARGYN...           1   \n",
       "89318  {\"cdr3\": \"CATSDTNAYKVIF\", \"cdr3_old\": \"CATSDTN...           1   \n",
       "89319  {\"cdr3\": \"CALSDTGANTGKLTF\", \"cdr3_old\": \"CALSD...           0   \n",
       "89320  {\"cdr3\": \"CALRGNNNNRIFF\", \"cdr3_old\": \"CALRGNN...           0   \n",
       "\n",
       "       web.method web.method.seq web.cdr3fix.nc web.cdr3fix.unmp  \n",
       "0            sort         sanger             no               no  \n",
       "1            sort         sanger             no               no  \n",
       "2            sort         sanger             no               no  \n",
       "3            sort         sanger             no               no  \n",
       "4            sort         sanger             no               no  \n",
       "...           ...            ...            ...              ...  \n",
       "89316        sort         sanger             no               no  \n",
       "89317        sort         sanger             no               no  \n",
       "89318        sort         sanger             no               no  \n",
       "89319        sort         sanger             no               no  \n",
       "89320        sort         sanger             no               no  \n",
       "\n",
       "[89321 rows x 21 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# read VDjdb dataset\n",
    "VDJdb = pd.read_csv(\"vdjdb.txt\", sep = \"\\t\")\n",
    "display(VDJdb)\n",
    "#VDJdb.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "1e6ad5fc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>cdr3_a_aa</th>\n",
       "      <th>v_a_gene</th>\n",
       "      <th>j_a_gene</th>\n",
       "      <th>species</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [gene, cdr3_a_aa, v_a_gene, j_a_gene, species, count]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>cdr3_a_aa</th>\n",
       "      <th>v_a_gene</th>\n",
       "      <th>j_a_gene</th>\n",
       "      <th>species</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TRA</td>\n",
       "      <td>CIVRAPGRADMRF</td>\n",
       "      <td>TRAV26-1*01</td>\n",
       "      <td>TRAJ43*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TRA</td>\n",
       "      <td>CAVPSGAGSYQLTF</td>\n",
       "      <td>TRAV20*01</td>\n",
       "      <td>TRAJ28*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>TRA</td>\n",
       "      <td>CAVKASGSRLT</td>\n",
       "      <td>TRAV2*01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>TRA</td>\n",
       "      <td>CAYRPPGTYKYIF</td>\n",
       "      <td>TRAV38-2/DV8*01</td>\n",
       "      <td>TRAJ40*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>TRA</td>\n",
       "      <td>CIVRAPGRADMRF</td>\n",
       "      <td>TRAV26-1*01</td>\n",
       "      <td>TRAJ43*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89068</th>\n",
       "      <td>TRA</td>\n",
       "      <td>CAVYNQGGKLIF</td>\n",
       "      <td>TRAV8-1*01</td>\n",
       "      <td>TRAJ23*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89070</th>\n",
       "      <td>TRA</td>\n",
       "      <td>CAVRALYNFNKFYF</td>\n",
       "      <td>TRAV3*01</td>\n",
       "      <td>TRAJ21*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89072</th>\n",
       "      <td>TRA</td>\n",
       "      <td>CAVESYNTDKLIF</td>\n",
       "      <td>TRAV41*01</td>\n",
       "      <td>TRAJ34*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89075</th>\n",
       "      <td>TRA</td>\n",
       "      <td>CLVGENSGGFKTIF</td>\n",
       "      <td>TRAV4*01</td>\n",
       "      <td>TRAJ9*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89076</th>\n",
       "      <td>TRA</td>\n",
       "      <td>CAMREGQGNAGNMLTF</td>\n",
       "      <td>TRAV14/DV4*01</td>\n",
       "      <td>TRAJ39*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>34188 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      gene         cdr3_a_aa         v_a_gene   j_a_gene      species  count\n",
       "0      TRA     CIVRAPGRADMRF      TRAV26-1*01  TRAJ43*01  HomoSapiens      1\n",
       "3      TRA    CAVPSGAGSYQLTF        TRAV20*01  TRAJ28*01  HomoSapiens      1\n",
       "5      TRA       CAVKASGSRLT         TRAV2*01        NaN  HomoSapiens      1\n",
       "7      TRA     CAYRPPGTYKYIF  TRAV38-2/DV8*01  TRAJ40*01  HomoSapiens      1\n",
       "9      TRA     CIVRAPGRADMRF      TRAV26-1*01  TRAJ43*01  HomoSapiens      1\n",
       "...    ...               ...              ...        ...          ...    ...\n",
       "89068  TRA      CAVYNQGGKLIF       TRAV8-1*01  TRAJ23*01  HomoSapiens      1\n",
       "89070  TRA    CAVRALYNFNKFYF         TRAV3*01  TRAJ21*01  HomoSapiens      1\n",
       "89072  TRA     CAVESYNTDKLIF        TRAV41*01  TRAJ34*01  HomoSapiens      1\n",
       "89075  TRA    CLVGENSGGFKTIF         TRAV4*01   TRAJ9*01  HomoSapiens      1\n",
       "89076  TRA  CAMREGQGNAGNMLTF    TRAV14/DV4*01  TRAJ39*01  HomoSapiens      1\n",
       "\n",
       "[34188 rows x 6 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>cdr3_b_aa</th>\n",
       "      <th>v_b_gene</th>\n",
       "      <th>j_b_gene</th>\n",
       "      <th>species</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TRB</td>\n",
       "      <td>CASSYLPGQGDHYSNQPQHF</td>\n",
       "      <td>TRBV13*01</td>\n",
       "      <td>TRBJ1-5*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TRB</td>\n",
       "      <td>CASSFEAGQGFFSNQPQHF</td>\n",
       "      <td>TRBV13*01</td>\n",
       "      <td>TRBJ1-5*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TRB</td>\n",
       "      <td>CASSFEPGQGFYSNQPQHF</td>\n",
       "      <td>TRBV13*01</td>\n",
       "      <td>TRBJ1-5*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>TRB</td>\n",
       "      <td>CASSYEPGQVSHYSNQPQHF</td>\n",
       "      <td>TRBV13*01</td>\n",
       "      <td>TRBJ1-5*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>TRB</td>\n",
       "      <td>CASSALASLNEQFF</td>\n",
       "      <td>TRBV14*01</td>\n",
       "      <td>TRBJ2-1*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89069</th>\n",
       "      <td>TRB</td>\n",
       "      <td>CASSALVAFSTDTQYF</td>\n",
       "      <td>TRBV6-4*01</td>\n",
       "      <td>TRBJ2-3*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89071</th>\n",
       "      <td>TRB</td>\n",
       "      <td>CASSSLAGEVNEQFF</td>\n",
       "      <td>TRBV7-2*01</td>\n",
       "      <td>TRBJ2-1*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89073</th>\n",
       "      <td>TRB</td>\n",
       "      <td>CASSSYTQRGLDYTGELFF</td>\n",
       "      <td>TRBV27*01</td>\n",
       "      <td>TRBJ2-2*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89074</th>\n",
       "      <td>TRB</td>\n",
       "      <td>CSVEDPMGVGTEAFF</td>\n",
       "      <td>TRBV29-1*01</td>\n",
       "      <td>TRBJ1-1*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89077</th>\n",
       "      <td>TRB</td>\n",
       "      <td>CASSQDYWGGGADTQYF</td>\n",
       "      <td>TRBV4-1*01</td>\n",
       "      <td>TRBJ2-3*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>47678 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      gene             cdr3_b_aa     v_b_gene    j_b_gene      species  count\n",
       "1      TRB  CASSYLPGQGDHYSNQPQHF    TRBV13*01  TRBJ1-5*01  HomoSapiens      1\n",
       "2      TRB   CASSFEAGQGFFSNQPQHF    TRBV13*01  TRBJ1-5*01  HomoSapiens      1\n",
       "4      TRB   CASSFEPGQGFYSNQPQHF    TRBV13*01  TRBJ1-5*01  HomoSapiens      1\n",
       "6      TRB  CASSYEPGQVSHYSNQPQHF    TRBV13*01  TRBJ1-5*01  HomoSapiens      1\n",
       "8      TRB        CASSALASLNEQFF    TRBV14*01  TRBJ2-1*01  HomoSapiens      1\n",
       "...    ...                   ...          ...         ...          ...    ...\n",
       "89069  TRB      CASSALVAFSTDTQYF   TRBV6-4*01  TRBJ2-3*01  HomoSapiens      1\n",
       "89071  TRB       CASSSLAGEVNEQFF   TRBV7-2*01  TRBJ2-1*01  HomoSapiens      1\n",
       "89073  TRB   CASSSYTQRGLDYTGELFF    TRBV27*01  TRBJ2-2*01  HomoSapiens      1\n",
       "89074  TRB       CSVEDPMGVGTEAFF  TRBV29-1*01  TRBJ1-1*01  HomoSapiens      1\n",
       "89077  TRB     CASSQDYWGGGADTQYF   TRBV4-1*01  TRBJ2-3*01  HomoSapiens      1\n",
       "\n",
       "[47678 rows x 6 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>cdr3_a_aa</th>\n",
       "      <th>v_a_gene</th>\n",
       "      <th>j_a_gene</th>\n",
       "      <th>cdr3_b_aa</th>\n",
       "      <th>v_b_gene</th>\n",
       "      <th>j_b_gene</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>CIVRAPGRADMRF</td>\n",
       "      <td>TRAV26-1*01</td>\n",
       "      <td>TRAJ43*01</td>\n",
       "      <td>CASSYLPGQGDHYSNQPQHF</td>\n",
       "      <td>TRBV13*01</td>\n",
       "      <td>TRBJ1-5*01</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>CAVPSGAGSYQLTF</td>\n",
       "      <td>TRAV20*01</td>\n",
       "      <td>TRAJ28*01</td>\n",
       "      <td>CASSFEPGQGFYSNQPQHF</td>\n",
       "      <td>TRBV13*01</td>\n",
       "      <td>TRBJ1-5*01</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>CAVKASGSRLT</td>\n",
       "      <td>TRAV2*01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CASSYEPGQVSHYSNQPQHF</td>\n",
       "      <td>TRBV13*01</td>\n",
       "      <td>TRBJ1-5*01</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>CAYRPPGTYKYIF</td>\n",
       "      <td>TRAV38-2/DV8*01</td>\n",
       "      <td>TRAJ40*01</td>\n",
       "      <td>CASSALASLNEQFF</td>\n",
       "      <td>TRBV14*01</td>\n",
       "      <td>TRBJ2-1*01</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>CIVRAPGRADMRF</td>\n",
       "      <td>TRAV26-1*01</td>\n",
       "      <td>TRAJ43*01</td>\n",
       "      <td>CASSYLPGQGDHYSNQPQHF</td>\n",
       "      <td>TRBV13*01</td>\n",
       "      <td>TRBJ1-5*01</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59805</th>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>CAFLNAGGTSYGKLTF</td>\n",
       "      <td>TRAV38-1*01</td>\n",
       "      <td>TRAJ52*01</td>\n",
       "      <td>CASSWNGLAGDEQFF</td>\n",
       "      <td>TRBV6-2*01</td>\n",
       "      <td>TRBJ2-1*01</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59806</th>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>CAVYNQGGKLIF</td>\n",
       "      <td>TRAV8-1*01</td>\n",
       "      <td>TRAJ23*01</td>\n",
       "      <td>CASSALVAFSTDTQYF</td>\n",
       "      <td>TRBV6-4*01</td>\n",
       "      <td>TRBJ2-3*01</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59807</th>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>CAVRALYNFNKFYF</td>\n",
       "      <td>TRAV3*01</td>\n",
       "      <td>TRAJ21*01</td>\n",
       "      <td>CASSSLAGEVNEQFF</td>\n",
       "      <td>TRBV7-2*01</td>\n",
       "      <td>TRBJ2-1*01</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59808</th>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>CAVESYNTDKLIF</td>\n",
       "      <td>TRAV41*01</td>\n",
       "      <td>TRAJ34*01</td>\n",
       "      <td>CASSSYTQRGLDYTGELFF</td>\n",
       "      <td>TRBV27*01</td>\n",
       "      <td>TRBJ2-2*01</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59811</th>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>CAMREGQGNAGNMLTF</td>\n",
       "      <td>TRAV14/DV4*01</td>\n",
       "      <td>TRAJ39*01</td>\n",
       "      <td>CASSQDYWGGGADTQYF</td>\n",
       "      <td>TRBV4-1*01</td>\n",
       "      <td>TRBJ2-3*01</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>27006 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           species         cdr3_a_aa         v_a_gene   j_a_gene  \\\n",
       "0      HomoSapiens     CIVRAPGRADMRF      TRAV26-1*01  TRAJ43*01   \n",
       "2      HomoSapiens    CAVPSGAGSYQLTF        TRAV20*01  TRAJ28*01   \n",
       "3      HomoSapiens       CAVKASGSRLT         TRAV2*01        NaN   \n",
       "4      HomoSapiens     CAYRPPGTYKYIF  TRAV38-2/DV8*01  TRAJ40*01   \n",
       "5      HomoSapiens     CIVRAPGRADMRF      TRAV26-1*01  TRAJ43*01   \n",
       "...            ...               ...              ...        ...   \n",
       "59805  HomoSapiens  CAFLNAGGTSYGKLTF      TRAV38-1*01  TRAJ52*01   \n",
       "59806  HomoSapiens      CAVYNQGGKLIF       TRAV8-1*01  TRAJ23*01   \n",
       "59807  HomoSapiens    CAVRALYNFNKFYF         TRAV3*01  TRAJ21*01   \n",
       "59808  HomoSapiens     CAVESYNTDKLIF        TRAV41*01  TRAJ34*01   \n",
       "59811  HomoSapiens  CAMREGQGNAGNMLTF    TRAV14/DV4*01  TRAJ39*01   \n",
       "\n",
       "                  cdr3_b_aa    v_b_gene    j_b_gene  count  \n",
       "0      CASSYLPGQGDHYSNQPQHF   TRBV13*01  TRBJ1-5*01      1  \n",
       "2       CASSFEPGQGFYSNQPQHF   TRBV13*01  TRBJ1-5*01      1  \n",
       "3      CASSYEPGQVSHYSNQPQHF   TRBV13*01  TRBJ1-5*01      1  \n",
       "4            CASSALASLNEQFF   TRBV14*01  TRBJ2-1*01      1  \n",
       "5      CASSYLPGQGDHYSNQPQHF   TRBV13*01  TRBJ1-5*01      1  \n",
       "...                     ...         ...         ...    ...  \n",
       "59805       CASSWNGLAGDEQFF  TRBV6-2*01  TRBJ2-1*01      1  \n",
       "59806      CASSALVAFSTDTQYF  TRBV6-4*01  TRBJ2-3*01      1  \n",
       "59807       CASSSLAGEVNEQFF  TRBV7-2*01  TRBJ2-1*01      1  \n",
       "59808   CASSSYTQRGLDYTGELFF   TRBV27*01  TRBJ2-2*01      1  \n",
       "59811     CASSQDYWGGGADTQYF  TRBV4-1*01  TRBJ2-3*01      1  \n",
       "\n",
       "[27006 rows x 8 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# To get the data structure as TCRdist requireabs\n",
    "# To process the data frame, separate the TRA and TRB. and select all rows of HomoSpaiens\n",
    "VDJdb_TRA_TRB = VDJdb[['gene', 'cdr3', 'v.segm', 'j.segm', 'species']]\n",
    "VDJdb_full = VDJdb_full[[\"species\", \"cdr3.alpha\", \"v.alpha\", \"j.alpha\", \"cdr3.beta\", \"v.beta\", \"j.beta\"]]\n",
    "\n",
    "\n",
    "VDJdb_TRA = VDJdb_TRA_TRB[ VDJdb_TRA_TRB['gene'] == \"TRA\"]\n",
    "VDJdb_TRB = VDJdb_TRA_TRB[ VDJdb_TRA_TRB['gene'] == \"TRB\"]\n",
    "\n",
    "VDJdb_TRA_human = VDJdb_TRA[ VDJdb_TRA['species'] == \"HomoSapiens\"]\n",
    "VDJdb_TRB_human = VDJdb_TRB[ VDJdb_TRB['species'] == \"HomoSapiens\"]\n",
    "VDJdb_TRA_mouse = VDJdb_TRA[ VDJdb_TRA['species'] == \"MusMusculus\"]\n",
    "VDJdb_TRB_mouse = VDJdb_TRB[ VDJdb_TRB['species'] == \"MusMusculus\"]\n",
    "VDJdb_TRA_TRB_human = VDJdb_full[ VDJdb_full['species'] == \"HomoSapiens\"]\n",
    "VDJdb_TRA_TRB_mouse = VDJdb_full[ VDJdb_full['species'] == \"MusMusculus\"]\n",
    "\n",
    "# Rename the columnms to satisfied the requirement of TCRdist.\n",
    "# For Human\n",
    "VDJdb_TRA_human_fix = VDJdb_TRA_human.rename(columns = {'cdr3': 'cdr3_a_aa', 'v.segm': 'v_a_gene', 'j.segm': 'j_a_gene'})\n",
    "VDJdb_TRB_human_fix = VDJdb_TRB_human.rename(columns = {'cdr3': 'cdr3_b_aa', 'v.segm': 'v_b_gene', 'j.segm': 'j_b_gene'})\n",
    "VDJdb_TRA_TRB_human_fix = VDJdb_TRA_TRB_human.rename(columns = {'cdr3.alpha': 'cdr3_a_aa', 'v.alpha': 'v_a_gene', \n",
    "                                                                'j.alpha':'j_a_gene', 'cdr3.beta': 'cdr3_b_aa', \n",
    "                                                                'v.beta':'v_b_gene', 'j.beta':'j_b_gene'})\n",
    "# For Mouse\n",
    "VDJdb_TRA_mouse_fix = VDJdb_TRA_mouse.rename(columns = {'cdr3': 'cdr3_a_aa', 'v.segm': 'v_a_gene', 'j.segm': 'j_a_gene'})\n",
    "VDJdb_TRB_mouse_fix = VDJdb_TRB_mouse.rename(columns = {'cdr3': 'cdr3_b_aa', 'v.segm': 'v_b_gene', 'j.segm': 'j_b_gene'})\n",
    "VDJdb_TRA_TRB_mouse_fix = VDJdb_TRA_TRB_mouse.rename(columns = {'cdr3.alpha': 'cdr3_a_aa', 'v.alpha': 'v_a_gene', \n",
    "                                                                'j.alpha':'j_a_gene', 'cdr3.beta': 'cdr3_b_aa', \n",
    "                                                                'v.beta':'v_b_gene', 'j.beta':'j_b_gene'})\n",
    "\n",
    "\n",
    "# Add a column called count\n",
    "VDJdb_TRA_human_fix['count'] = 1\n",
    "VDJdb_TRB_human_fix['count'] = 1\n",
    "VDJdb_TRA_TRB_human_fix['count'] = 1\n",
    "VDJdb_TRA_mouse_fix['count'] = 1\n",
    "VDJdb_TRB_mouse_fix['count'] = 1\n",
    "VDJdb_TRA_TRB_mouse_fix['count'] = 1\n",
    "\n",
    "\n",
    "# Remove rows with NAN value\n",
    "# For humanabs\n",
    "VDJdb_TRA_human_fix = VDJdb_TRA_human_fix.dropna(subset = ['v_a_gene'])\n",
    "VDJdb_TRB_human_fix = VDJdb_TRB_human_fix.dropna(subset = ['v_b_gene'])\n",
    "VDJdb_TRA_TRB_human_fix = VDJdb_TRA_TRB_human_fix.dropna(subset = ['v_a_gene', 'cdr3_a_aa', 'cdr3_b_aa', 'v_b_gene'])\n",
    "\n",
    "# For mouse\n",
    "VDJdb_TRA_mouse_fix = VDJdb_TRA_mouse_fix.dropna(subset = ['v_a_gene'])\n",
    "VDJdb_TRB_mouse_fix = VDJdb_TRB_mouse_fix.dropna(subset = ['v_b_gene'])\n",
    "VDJdb_TRA_TRB_mouse_fix = VDJdb_TRA_TRB_mouse_fix.dropna(subset = ['v_a_gene', 'cdr3_a_aa', 'cdr3_b_aa', 'v_b_gene'])\n",
    "\n",
    "# Check the NAN row exist or not.\n",
    "check_NAN = VDJdb_TRA_human_fix[VDJdb_TRA_human_fix['v_a_gene'].isna()]\n",
    "display(check_NAN)\n",
    "\n",
    "# Display the data rows\n",
    "display(VDJdb_TRA_human_fix)\n",
    "display(VDJdb_TRB_human_fix)\n",
    "display(VDJdb_TRA_TRB_human_fix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "9b56db01",
   "metadata": {},
   "outputs": [],
   "source": [
    "# If you want to run the following code, please install tcrdist first.\n",
    "import pwseqdist as pw\n",
    "import pandas as pd\n",
    "from tcrdist.repertoire import TCRrep\n",
    "from tcrdist.vdjtools_funcs import import_vdjtools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "c2a14977",
   "metadata": {},
   "outputs": [],
   "source": [
    "# To process TRA chain to get distance matrix between TRA\n",
    "# For the df, which will contain 34188 rows,so the distance matrix should be 34188*34188 matrix.\n",
    "# My computer can not process the much big scale data. So I just select 10000*10000.\n",
    "# For the df, which will contain 34188 rows,so the distance matrix should be 34188*34188 matrix.\n",
    "# If any one of you can process the whole data. please do that.\n",
    "df_alpha = VDJdb_TRA_human_fix[0:10000]\n",
    "tr_alpha = TCRrep(cell_df = df_alpha,\n",
    "            organism = 'human',\n",
    "            chains = ['alpha'],\n",
    "            deduplicate = False,\n",
    "            db_file = \"alphabeta_gammadelta_db.tsv\",\n",
    "            compute_distances=False)\n",
    "\n",
    "df_beta = VDJdb_TRB_human_fix[0:10000]\n",
    "tr_beta = TCRrep(cell_df = df_beta,\n",
    "            organism = 'human',\n",
    "            chains = ['beta'],\n",
    "            deduplicate = False,\n",
    "            db_file = \"alphabeta_gammadelta_db.tsv\",\n",
    "            compute_distances=False)\n",
    "\n",
    "df_whole = VDJdb_TRA_TRB_human_fix[0:10000]\n",
    "tr_whole = TCRrep(cell_df = df_whole,\n",
    "            organism = 'human',\n",
    "            chains = ['alpha', 'beta'],\n",
    "            deduplicate = False,\n",
    "            db_file = \"alphabeta_gammadelta_db.tsv\",\n",
    "            compute_distances=False)\n",
    "\n",
    "\n",
    "#tr.compute_distances()\n",
    "#tr.cpus = 4\n",
    "#tr.compute_sparse_rect_distances(radius = 50, chunk_size = 80)\n",
    "#tr.rw_alpha\n",
    "#tr.pw_alpha"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "03078941",
   "metadata": {},
   "outputs": [],
   "source": [
    "# compute_sparse_rect_distances\n",
    "tr_alpha.cpus = 4\n",
    "tr_alpha.compute_sparse_rect_distances(radius = 50, chunk_size = 100)\n",
    "tr_alpha.rw_alpha"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "9902cb71",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "2b379f738e1146c3bd1db57d614afec7",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/100 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<10000x10000 sparse matrix of type '<class 'numpy.int16'>'\n",
       "\twith 214570 stored elements in Compressed Sparse Row format>"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# compute_sparse_rect_distances\n",
    "tr_beta.cpus = 4\n",
    "tr_beta.compute_sparse_rect_distances(radius = 50, chunk_size = 100)\n",
    "tr_beta.rw_beta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "30e8695a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# compute_sparse_rect_distances\n",
    "tr_beta.cpus = 4\n",
    "tr_beta.compute_sparse_rect_distances(radius = 50, chunk_size = 100)\n",
    "tr_alpha.rw_alpha\n",
    "tr_beta.rw_beta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "a70eb48c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  (0, 0)\t-1\n",
      "  (0, 4)\t-1\n",
      "  (0, 7)\t-1\n",
      "  (0, 10)\t-1\n",
      "  (1, 1)\t-1\n",
      "  (1, 5)\t-1\n",
      "  (1, 8)\t-1\n",
      "  (1, 11)\t-1\n",
      "  (2, 2)\t-1\n",
      "  (2, 6)\t-1\n",
      "  (2, 9)\t-1\n",
      "  (2, 12)\t-1\n",
      "  (3, 3)\t-1\n",
      "  (4, 0)\t-1\n",
      "  (4, 4)\t-1\n",
      "  (4, 7)\t-1\n",
      "  (4, 10)\t-1\n",
      "  (5, 1)\t-1\n",
      "  (5, 5)\t-1\n",
      "  (5, 8)\t-1\n",
      "  (5, 11)\t-1\n",
      "  (6, 2)\t-1\n",
      "  (6, 6)\t-1\n",
      "  (6, 9)\t-1\n",
      "  (6, 12)\t-1\n",
      "  :\t:\n",
      "  (498, 498)\t-1\n",
      "  (499, 63)\t44\n",
      "  (499, 64)\t47\n",
      "  (499, 65)\t47\n",
      "  (499, 66)\t47\n",
      "  (499, 369)\t24\n",
      "  (499, 385)\t24\n",
      "  (499, 388)\t24\n",
      "  (499, 392)\t44\n",
      "  (499, 397)\t24\n",
      "  (499, 398)\t44\n",
      "  (499, 403)\t24\n",
      "  (499, 411)\t44\n",
      "  (499, 413)\t47\n",
      "  (499, 414)\t24\n",
      "  (499, 417)\t44\n",
      "  (499, 433)\t24\n",
      "  (499, 435)\t44\n",
      "  (499, 438)\t21\n",
      "  (499, 450)\t44\n",
      "  (499, 455)\t44\n",
      "  (499, 464)\t44\n",
      "  (499, 467)\t44\n",
      "  (499, 493)\t21\n",
      "  (499, 499)\t-1\n"
     ]
    }
   ],
   "source": [
    "# Print sparse_rect_distances\n",
    "print(tr_alpha.rw_alpha)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "81a4f3bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compute alpha chain distance matrix \n",
    "tr_alpha.compute_distances()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "48c324ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compute beta chain distance matrix \n",
    "tr_beta.compute_distances()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2b2a59de",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compute whole chain distance matrix \n",
    "tr_whole.compute_distances()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "8a6bf7c0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[  0, 138, 125, ..., 152, 150, 158],\n",
       "       [138,   0, 138, ..., 110, 118, 128],\n",
       "       [125, 138,   0, ..., 140, 120, 154],\n",
       "       ...,\n",
       "       [152, 110, 140, ...,   0, 133,  80],\n",
       "       [150, 118, 120, ..., 133,   0, 144],\n",
       "       [158, 128, 154, ...,  80, 144,   0]], dtype=int16)"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Print distance matrix\n",
    "tr_alpha.pw_alpha"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0f03a383",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
